### PR TITLE
Add date and datetime timestamp providers

### PR DESCRIPTION
**Pull request by lubaskinc0de**
#281 

Tasks

- [x] Providers
- [x] Docs
- [x] Tests

```python
@dataclass
class Book:
    title: str
    price: int
    created_at: date


book = Book("How to be rich.", price=100, created_at=datetime.now())
retort = Retort(
    recipe=[
        DatetimeTimestampProvider(),
        DateTimestampProvider(),
    ]
)

print(retort.dump(book))
data = {'title': 'How to be rich.', 'price': 100, 'created_at': 1717766584.128659}
print(retort.load(data, Book))
```

### CODE REVIEW

File: src/adaptix/_internal/morphing/concrete_provider.py [hunk 1]
```diff
@@ -75,6 +75,51 @@ def datetime_format_dumper(data: datetime):
         return datetime_format_dumper
 
 
+@dataclass
+@for_predicate(datetime)
+class DatetimeTimestampProvider(LoaderProvider, DumperProvider):
+    tz: timezone = timezone.utc
```
Comment (daler-sz) [inline src/adaptix/_internal/morphing/concrete_provider.py:81]: Wouldn't be it better, if timezone would be received in `__init__`?

Comment (lubaskinc0de) [reply src/adaptix/_internal/morphing/concrete_provider.py:81]: it's dataclass what means less boilerplate and it's consistent with DatetimeFormatProvider that also receive format parameter through dataclass

Comment (daler-sz) [reply src/adaptix/_internal/morphing/concrete_provider.py:81]: Oh didn't notice that it is daraclass, my bad

**Comment by lubaskinc0de** [comment]
This is also closes #281

File: docs/changelog/fragments/286.bugfix.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Add public api for :func:`.datetime_format_provider`.
+`#286 <https://github.com/reagento/adaptix/issues/286>`_
```
Comment (daler-sz) [inline docs/changelog/fragments/286.bugfix.rst:2]: I don't think that docs changes mentions are appropriete for changelog

File: docs/loading-and-dumping/specific-types-behavior.rst [hunk 3]
```diff
@@ -109,7 +109,13 @@ dumper serialize value via ``__str__`` method.
 date, time and datetime
 '''''''''''''''''''''''''''
 
-Value is represented as an isoformat string.
+By default value is represented as an isoformat string.
+
+But you can override this behaviour with external providers:
```
Comment (daler-sz) [inline docs/loading-and-dumping/specific-types-behavior.rst:114]: "external" means "from another place", "third party", so I would chosen another term, or just "other (providers)"

File: src/adaptix/_internal/morphing/concrete_provider.py [hunk 4]
```diff
@@ -75,6 +75,57 @@ def datetime_format_dumper(data: datetime):
...
+@for_predicate(datetime)
+class DatetimeTimestampProvider(LoaderProvider, DumperProvider):
+    tz: typing.Optional[timezone] = timezone.utc
+
+    def _provide_loader(self, mediator: Mediator, request: LoaderRequest) -> Loader:
+        tz = self.tz
+
+        def datetime_timestamp_loader(data):
+            try:
+                if data is None:
```
Comment (Eclips4) [inline src/adaptix/_internal/morphing/concrete_provider.py:88]: Let's add some clarification here about why this guard is exist
```suggestion
                # Pure-Python implenentation and C-extension implementation
                # of datetime.date.fromtimestamp module works differently with a None arg.
                # See https://github.com/python/cpython/issues/120268 for more details.
                if data is None:
```

File: src/adaptix/_internal/morphing/concrete_provider.py [hunk 5]
```diff
@@ -75,6 +75,57 @@ def datetime_format_dumper(data: datetime):
...
+                    raise TypeLoadError(float, data)
+
+                return datetime.fromtimestamp(data, tz=tz)
+            except TypeError:
+                raise TypeLoadError(float, data)
+            except OverflowError:
+                raise ValueLoadError(
+                    "Timestamp is out of the range of values supported by the platform",
```
Comment (daler-sz) [inline src/adaptix/_internal/morphing/concrete_provider.py:96]: This error message is for API clients, so "by platform" is reduntant here. It should be just "... of supported values"

File: src/adaptix/_internal/morphing/facade/provider.py [hunk 6]
```diff
@@ -439,3 +441,36 @@ def default_dict(pred: Pred, default_factory: Callable) -> Provider:
...
+
+def datetime_timestamp_provider(pred: Pred, tz: Optional[timezone] = None) -> Provider:
+    """Provider that can load/dump datetime object from/to UNIX timestamp.
+
+    :param pred: Predicate specifying where the provider should be used.
+        See :ref:`predicate-system` for details.
+    :param tz: tz parameter which will be passed to the datetime.fromtimestamp method.
+    """
+
+    provider = DatetimeTimestampProvider(tz) if tz else DatetimeTimestampProvider()
```
Comment (daler-sz) [inline src/adaptix/_internal/morphing/facade/provider.py:454]: why not just use `timezone.utc` as the default value of tz param?

File: tests/unit/morphing/test_concrete_provider.py [hunk 7]
```diff
@@ -129,6 +124,79 @@ def test_datetime_format_provider(strict_coercion, debug_trail):
...
+    assert loader(ts) == today
+
+    check_any_dt(loader, float, invalid_objects=(
+        None,
+        datetime(2011, 11, 4, 0, 0),
+        date(2019, 12, 4),
+        time(4, 23, 1),
+    ))
+
+    overflow_ts = ts + 100 ** 10
```
Comment (daler-sz) [inline tests/unit/morphing/test_concrete_provider.py:192]: I think float("inf") is more suitable for this case

File: src/adaptix/_internal/morphing/concrete_provider.py [hunk 8]
```diff
@@ -93,7 +93,7 @@ def datetime_timestamp_loader(data):
                 raise TypeLoadError(float, data)
             except OverflowError:
                 raise ValueLoadError(
-                    "Timestamp is out of the range of values supported by the platform",
+                    "Timestamp is out of the range of values supported",
```
Comment (daler-sz) [inline src/adaptix/_internal/morphing/concrete_provider.py:96]: *supported values

File: src/adaptix/_internal/morphing/concrete_provider.py [hunk 9]
```diff
@@ -75,6 +75,57 @@ def datetime_format_dumper(data: datetime):
...
+                    raise TypeLoadError(float, data)
+
+                return date.fromtimestamp(data)  # noqa: DTZ012
+            except TypeError:
+                raise TypeLoadError(float, data)
+            except OverflowError:
+                raise ValueLoadError(
+                    "Timestamp is out of the range of supported values",
+                    data,
+                )
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/concrete_provider.py:124]: Most of library users use interpreter version that does not need this extra checking.

So, we need to add extra validation if this snippet returns ``True``
```python
try:
    import _pydatetime
except ImportError:
     return False
else:
     return datetime is _pydatetime
```

File: src/adaptix/_internal/morphing/facade/provider.py [hunk 10]
```diff
@@ -439,3 +441,35 @@ def default_dict(pred: Pred, default_factory: Callable) -> Provider:
...
+
+
+def date_timestamp_provider(pred: Pred) -> Provider:
+    """Provider that can load date object from UNIX timestamp.
+    Note that date objects can`t be dumped to the UNIX timestamp
+
+    :param pred: Predicate specifying where the provider should be used.
+        See :ref:`predicate-system` for details.
+    """
+    return bound(pred, DateTimestampProvider())
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/facade/provider.py:474]: This names are inconsistent with other names of public factories.
I think that ``datetime_by_timestamp``, ``datetime_by_format``, ``date_by_timestamp`` will look better.

File: src/adaptix/_internal/morphing/facade/provider.py [hunk 11]
```diff
@@ -439,3 +441,35 @@ def default_dict(pred: Pred, default_factory: Callable) -> Provider:
...
+
+def datetime_timestamp_provider(pred: Pred, tz: Optional[timezone] = None) -> Provider:
+    """Provider that can load/dump datetime object from/to UNIX timestamp.
+
+    :param pred: Predicate specifying where the provider should be used.
+        See :ref:`predicate-system` for details.
+    :param tz: tz parameter which will be passed to the datetime.fromtimestamp method.
+    """
+
+    return bound(pred, DatetimeTimestampProvider(tz))
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/facade/provider.py:454]: Naive datetimes consider harmful, default timezone should be UTC

File: tests/unit/morphing/test_concrete_provider.py [hunk 12]
```diff
@@ -12,31 +12,26 @@
...
+    10,
+    datetime(2011, 11, 4, 0, 0),
+    date(2019, 12, 4),
+    time(4, 23, 1),
+)):
+    for obj in invalid_objects:
+        raises_exc(
+            TypeLoadError(expected_type, obj),
+            lambda: loader(obj),  # noqa: B023
+        )
```
Comment (zhPavel) [inline tests/unit/morphing/test_concrete_provider.py:34]: I think that we should use pytests parametrization instrad of manual looping

File: tests/unit/morphing/test_concrete_provider.py [hunk 13]
```diff
@@ -129,6 +124,79 @@ def test_datetime_format_provider(strict_coercion, debug_trail):
...
 
 
+@pytest.mark.parametrize(
+    "tz",
+    [
+        None,
+        timezone(timedelta(hours=3)),
+    ],
+)
+def test_datetime_timestamp_provider(strict_coercion, debug_trail, tz):
```
Comment (zhPavel) [inline tests/unit/morphing/test_concrete_provider.py:134]: Need test for ``float('nan')``

File: tests/unit/morphing/test_concrete_provider.py [hunk 14]
```diff
@@ -129,6 +124,79 @@ def test_datetime_format_provider(strict_coercion, debug_trail):
...
+        strict_coercion=strict_coercion,
+        debug_trail=debug_trail,
+        recipe=[
+            DatetimeTimestampProvider(tz=tz),
+        ],
+    )
+
+    loader = retort.get_loader(datetime)
+
+    dt = datetime.now(tz=tz)
```
Comment (zhPavel) [inline tests/unit/morphing/test_concrete_provider.py:145]: I think that all tests have to use containt values

**Review by zhPavel — CHANGES_REQUESTED** [review]

File: src/adaptix/_internal/morphing/concrete_provider.py (hunk 4 above)
Comment (Eclips4) [reply src/adaptix/_internal/morphing/concrete_provider.py:88]: > I think this is a good idea but `implenentation` contains typo

fixed

File: src/adaptix/__init__.py [hunk 15]
```diff
@@ -64,6 +67,9 @@
     "flag_by_member_names",
     "name_mapping",
     "default_dict",
+    "datetime_format_provider",
+    "date_timestamp_provider",
+    "datetime_timestamp_provider",
```
Comment (zhPavel) [inline src/adaptix/__init__.py:72]: The old names are used here

File: src/adaptix/_internal/morphing/concrete_provider.py [hunk 16]
```diff
@@ -75,6 +75,86 @@ def datetime_format_dumper(data: datetime):
...
+        except ImportError:
+            return False
+        else:
+            if datetime is _pydatetime:
+                return True
+
+        return False
+
+    def _provide_loader(self, mediator: Mediator, request: LoaderRequest) -> Loader:
+        is_pydatetime = self._is_pydatetime()
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/concrete_provider.py:122]: This variable looks redundant

File: src/adaptix/_internal/morphing/concrete_provider.py [hunk 17]
```diff
@@ -75,6 +75,86 @@ def datetime_format_dumper(data: datetime):
...
+                return date.fromtimestamp(data)  # noqa: DTZ012
+            except TypeError:
+                raise TypeLoadError(Union[int, float], data)
+            except OverflowError:
+                raise ValueLoadError(
+                    "Timestamp is out of the range of supported values",
+                    data,
+                )
+
+        return date_timestamp_loader if not is_pydatetime else pydate_timestamp_loader
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/concrete_provider.py:155]: Why is a negative condition used?

File: src/adaptix/_internal/morphing/facade/provider.py [hunk 18]
```diff
@@ -439,3 +441,35 @@ def default_dict(pred: Pred, default_factory: Callable) -> Provider:
...
+
+
+def date_by_timestamp(pred: Pred) -> Provider:
+    """Provider that can load date object from UNIX timestamp.
+    Note that date objects can`t be dumped to the UNIX timestamp
+
+    :param pred: Predicate specifying where the provider should be used.
+        See :ref:`predicate-system` for details.
+    """
+    return bound(pred, DateTimestampProvider())
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/facade/provider.py:474]: I think that a lot of people want to apply this rule to all datetimes or dates, so we need zero arguments signature.
And second argument will become keyword-only.

File: tests/unit/morphing/test_concrete_provider.py [hunk 19]
```diff
@@ -10,34 +10,68 @@
...
+        ), date, [DateTimestampProvider()]),
+    ],
+)
+def test_any_dt(
+    expected_type: typing.Type,
+    invalid_objects: typing.Tuple[typing.Any, ...],
+    loader_type: typing.Type,
+    extra_providers: typing.List[Provider],
+    strict_coercion,
+    debug_trail,
```
Comment (zhPavel) [inline tests/unit/morphing/test_concrete_provider.py:50]: Accessing via the typing module should only be used for types that are not present in all versions of Python

File: tests/unit/morphing/test_concrete_provider.py [hunk 20]
```diff
@@ -10,34 +10,68 @@
...
+            time(4, 23, 1),
+        ), datetime, [DatetimeTimestampProvider(tz=timezone.utc)]),
+        (Union[float, int], (
+            None,
+            datetime(2011, 11, 4, 0, 0),
+            date(2019, 12, 4),
+            time(4, 23, 1),
+        ), date, [DateTimestampProvider()]),
+    ],
+)
```
Comment (zhPavel) [inline tests/unit/morphing/test_concrete_provider.py:43]: We can express absent of extra providers via empty list

Comment (zhPavel) [inline tests/unit/morphing/test_concrete_provider.py:43]: There are two set of invalid input data. I think that two constants will looks better than copied literals and None

File: tests/unit/morphing/test_concrete_provider.py [hunk 21]
```diff
@@ -10,34 +10,68 @@
...
-    )
-    raises_exc(
-        TypeLoadError(str, date(2019, 12, 4)),
-        lambda: loader(date(2019, 12, 4)),
-    )
-    raises_exc(
-        TypeLoadError(str, time(4, 23, 1)),
-        lambda: loader(time(4, 23, 1)),
+@pytest.mark.parametrize(
+    ["expected_type", "invalid_objects", "loader_type", "extra_providers"],
```
Comment (zhPavel) [inline tests/unit/morphing/test_concrete_provider.py:24]: ``loader_type`` sounds like "type of loader". You can use name ``tp`` to show which type is tested

File: tests/unit/morphing/test_concrete_provider.py [hunk 22]
```diff
@@ -10,34 +10,68 @@
...
+        ), datetime, [DatetimeTimestampProvider(tz=timezone.utc)]),
+        (Union[float, int], (
+            None,
+            datetime(2011, 11, 4, 0, 0),
+            date(2019, 12, 4),
+            time(4, 23, 1),
+        ), date, [DateTimestampProvider()]),
+    ],
+)
+def test_any_dt(
```
Comment (zhPavel) [inline tests/unit/morphing/test_concrete_provider.py:44]: The name of test says nothing about what is it testing

File: tests/unit/morphing/test_concrete_provider.py [hunk 23]
```diff
@@ -10,34 +10,68 @@
...
+    expected_type: typing.Type,
+    invalid_objects: typing.Tuple[typing.Any, ...],
+    loader_type: typing.Type,
+    extra_providers: typing.List[Provider],
+    strict_coercion,
+    debug_trail,
+):
```
Comment (zhPavel) [inline tests/unit/morphing/test_concrete_provider.py:51]: I do not like that this test tries to check different providers.
It looks like we need 1 test for iso format, 1 test for timestamp and 1 test for custom format

File: src/adaptix/_internal/morphing/concrete_provider.py [hunk 24]
```diff
@@ -75,6 +75,86 @@ def datetime_format_dumper(data: datetime):
...
+
+    def _provide_dumper(self, mediator: Mediator, request: DumperRequest) -> Dumper:
+        def datetime_timestamp_dumper(data: datetime):
+            return data.timestamp()
+
+        return datetime_timestamp_dumper
+
+
+@for_predicate(date)
+class DateTimestampProvider(LoaderProvider):
```
Comment (zhPavel) [inline src/adaptix/_internal/morphing/concrete_provider.py:109]: This provider requires dumper

File: tests/unit/morphing/test_concrete_provider.py [hunk 25]
```diff
@@ -129,6 +157,77 @@ def test_datetime_format_provider(strict_coercion, debug_trail):
...
+
+    raises_exc(
+        ValueLoadError("Timestamp is out of the range of supported values", overflow_ts),
+        lambda: loader(overflow_ts),
+    )
+
+    raises_exc(
+        ValueLoadError("Unexpected value", nan),
+        lambda: loader(nan),
+    )
```
Comment (zhPavel) [inline tests/unit/morphing/test_concrete_provider.py:194]: When you create different values ​​at the beginning, and then test them in a row, you get the impression that these values ​​or tests are somehow logically connected. However, these are completely different errors

**Review by zhPavel — CHANGES_REQUESTED** [review]

File: src/adaptix/_internal/morphing/facade/provider.py (hunk 18 above)
Comment (zhPavel) [reply src/adaptix/_internal/morphing/facade/provider.py]: To create zero arguments signature, pred argument have to default value `P.ANY`

**Pull request merged**